# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/notification/NotificationUtils.cpp
# Commit: ed71fdc2d78d
# Full Hash: ed71fdc2d78dec085a3d153220bba4aeda89992b
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2025-03-05 04:28:59
# Regressor Bug: 1951379
# File Overlap Count: 1
# Description:
#   Bug 1951379 - Part 3: Add NotificationStorageEntry r=asuth
#   
#   This skips handling `vibrate` for now, as we don't ship it and it also complicates test setup.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D240213
# ==============================================================================

diff -r 61ab7888abc4 -r ed71fdc2d78d dom/notification/NotificationUtils.cpp
--- a/dom/notification/NotificationUtils.cpp	Tue Mar 04 18:00:17 2025 +0000
+++ b/dom/notification/NotificationUtils.cpp	Tue Mar 04 18:00:18 2025 +0000
@@ -166,8 +166,8 @@
                                  : NS_NOTIFICATION_STORAGE_CONTRACTID);
 }
 
-nsresult PersistNotification(nsIPrincipal* aPrincipal, const nsString& aId,
-                             const IPCNotificationOptions& aOptions,
+nsresult PersistNotification(nsIPrincipal* aPrincipal,
+                             const IPCNotification& aNotification,
                              const nsString& aScope) {
   nsCOMPtr<nsINotificationStorage> notificationStorage =
       GetNotificationStorage(aPrincipal->GetIsInPrivateBrowsing());
@@ -181,10 +181,10 @@
     return rv;
   }
 
-  rv = notificationStorage->Put(
-      origin, aId, aOptions.title(), GetEnumString(aOptions.dir()),
-      aOptions.lang(), aOptions.body(), aOptions.tag(), aOptions.icon(),
-      aOptions.dataSerialized(), aScope);
+  RefPtr<NotificationStorageEntry> entry =
+      new NotificationStorageEntry(aNotification);
+
+  rv = notificationStorage->Put(origin, entry, aScope);
 
   if (NS_FAILED(rv)) {
     return rv;
@@ -257,4 +257,81 @@
   return pushQuotaManager->NotificationForOriginClosed(origin.get());
 }
 
+NS_IMPL_ISUPPORTS(NotificationStorageEntry, nsINotificationStorageEntry)
+
+NS_IMETHODIMP NotificationStorageEntry::GetId(nsAString& aId) {
+  aId = mIPCNotification.id();
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetTitle(nsAString& aTitle) {
+  aTitle = mIPCNotification.options().title();
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetDir(nsACString& aDir) {
+  aDir = GetEnumString(mIPCNotification.options().dir());
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetLang(nsAString& aLang) {
+  aLang = mIPCNotification.options().lang();
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetBody(nsAString& aBody) {
+  aBody = mIPCNotification.options().body();
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetTag(nsAString& aTag) {
+  aTag = mIPCNotification.options().tag();
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetIcon(nsAString& aIcon) {
+  aIcon = mIPCNotification.options().icon();
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetRequireInteraction(
+    bool* aRequireInteraction) {
+  *aRequireInteraction = mIPCNotification.options().requireInteraction();
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetSilent(bool* aSilent) {
+  *aSilent = mIPCNotification.options().silent();
+  return NS_OK;
+}
+
+NS_IMETHODIMP NotificationStorageEntry::GetDataSerialized(
+    nsAString& aDataSerialized) {
+  aDataSerialized = mIPCNotification.options().dataSerialized();
+  return NS_OK;
+}
+
+Result<IPCNotification, nsresult> NotificationStorageEntry::ToIPC(
+    nsINotificationStorageEntry& aEntry) {
+  IPCNotification notification;
+  IPCNotificationOptions& options = notification.options();
+  MOZ_TRY(aEntry.GetId(notification.id()));
+  MOZ_TRY(aEntry.GetTitle(options.title()));
+
+  nsCString dir;
+  MOZ_TRY(aEntry.GetDir(dir));
+  options.dir() = StringToEnum<NotificationDirection>(dir).valueOr(
+      NotificationDirection::Auto);
+
+  MOZ_TRY(aEntry.GetLang(options.lang()));
+  MOZ_TRY(aEntry.GetBody(options.body()));
+  MOZ_TRY(aEntry.GetTag(options.tag()));
+  MOZ_TRY(aEntry.GetIcon(options.icon()));
+  MOZ_TRY(aEntry.GetRequireInteraction(&options.requireInteraction()));
+  MOZ_TRY(aEntry.GetSilent(&options.silent()));
+  MOZ_TRY(aEntry.GetDataSerialized(options.dataSerialized()));
+
+  return notification;
+}
+
 }  // namespace mozilla::dom::notification